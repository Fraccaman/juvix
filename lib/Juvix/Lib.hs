{-# LANGUAGE EmptyDataDecls         #-}
{-# LANGUAGE ExplicitForAll         #-}
{-# LANGUAGE FlexibleInstances      #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE MagicHash              #-}
{-# LANGUAGE MultiParamTypeClasses  #-}
{-# LANGUAGE RebindableSyntax       #-}
{-# LANGUAGE UnicodeSyntax          #-}

{-  Prelude-equivalent for Juvix-compiled programs.
    Covers most of the Haskell Prelude API interface with identical semantics (functions with the same name do the same thing).
    This file is mostly a lie; these functions are replaced by the transpiler with Michelson builtins.
    The intent of this library is to be "batteries-included" - so additional common functionality should be added as necessary.
    Subject to API instability until Michelson is finalized.  -}

module Juvix.Lib (

  {- Datatypes -}

  List,
  Set,
  Map,
  Signature,
  Key,
  Hash,
  Contract,
  Account,
  Either(..),
  Maybe(..),
  Pair,
  Unit,
  Tez,
  String,
  Int,
  Nat,
  Bool(..),

  {- General -}

  IO,
  Applicative,
  Functor,
  Monad,
  Monoid,
  pure,
  (<*>),
  fmap,
  (>>=),
  (>>),
  return,
  mappend,
  mempty,
  (<>),

  {- Rebindable Syntax -}

  ifThenElse,
  fromInteger,

  {- Booleans -}

  (&&),
  (||),
  otherwise,
  not,
  xor,

  {- Maybe -}

  fromMaybe,

  {- Comparision -}

  Ord,
  compare,
  max,
  min,
  (<),
  (<=),
  (>),
  (>=),

  {- Equality -}

  Eq,
  (==),
  (/=),

  {- Arithmetic -}

  Mul,
  Add,
  Div,
  Sub,
  Negate,
  (+),
  (-),
  (*),
  (/),
  negate,
  abs,

  {- Michelson Calls -}

  checkSignature,
  manager,
  defaultAccount,
  source,
  self,
  transferTokens,
  now,
  amount,
  balance,
  hash,
  stepsToQuota,
  fail,

  {- Misc Michelson -}

  emptySet,
  setMember,
  setUpdate,
  setSize,
  emptyMap,
  mapGet,
  mapMember,
  mapUpdate,
  mapSize

) where

{- These are only imported in order to support GHC's literal overloading. -}

import qualified Data.String (String)
import qualified GHC.Num     (Integer)

class Functor f where
  {-# MINIMAL fmap #-}

  fmap ∷ ∀ a b . (a → b) → f a → f b

  (<$) ∷ ∀ a b . a → f b → f a
  (<$) = fmap . const

infixl 4 <$

class (Functor f) ⇒ Applicative f where
  {-# MINIMAL pure, ((<*>) | liftA2) #-}

  pure ∷ ∀ a . a → f a

  (<*>) ∷ ∀ a b . f (a → b) → f a → f b
  (<*>) = liftA2 id

  liftA2 ∷ ∀ a b c . (a → b → c) → f a → f b → f c
  liftA2 f x y = f <$> x <*> y

  (*>) ∷ ∀ a b . f a → f b → f b
  (*>) x y = (id <$ x) <*> y

  (<*) ∷ ∀ a b . f a → f b → f a
  (<*) x y = liftA2 const x y

(<$>) ∷ Functor f ⇒ (a → b) → f a → f b
(<$>) = fmap

infixl 4 <$>

infixl 4 <*>

infixl 4 *>

infixl 4 <*

class (Applicative m) ⇒ Monad m where
  {-# MINIMAL (>>=) #-}

  (>>=) ∷ ∀ a b . m a → (a → m b) → m b

  (>>) ∷ ∀ a b . m a → m b → m b
  (>>) = (*>)

  return ∷ ∀ a . a → m a
  return = pure

infixl 1 >>=

infixl 1 >>

class Eq a where
  {-# MINIMAL (==) | (/=) #-}

  (==) ∷ a → a → Bool
  (==) x y = not (x /= y)
  {-# INLINABLE (==) #-}

  (/=) ∷ a → a → Bool
  (/=) x y = not (x == y)
  {-# INLINABLE (/=) #-}

infix 4 ==

infix 4 /=

class Monoid a where
  {-# MINIMAL mempty, mappend #-}

  mempty ∷ a

  mappend ∷ a → a → a

infixr 6 <>

(<>) ∷ ∀ a . (Monoid a) ⇒ a → a → a
(<>) = mappend

class (Eq a) ⇒ Ord a where
  {-# MINIMAL compare | (<=) #-}

  compare ∷ a → a → Ordering
  compare x y =
    if x == y then EQ
      else if x <= y then LT else GT

  (<) ∷ a → a → Bool
  (<) x y = compare x y == LT

  (<=) ∷ a → a → Bool
  (<=) x y = compare x y /= GT

  (>=) ∷ a → a → Bool
  (>=) x y = compare x y /= LT

  (>) ∷ a → a → Bool
  (>) x y = compare x y == GT

  max ∷ a → a → a
  max x y = if x >= y then x else y

  min ∷ a → a → a
  min x y = if x <= y then x else y

ifThenElse ∷ ∀ a . Bool → a → a → a
ifThenElse True  t _ = t
ifThenElse False _ f = f

data Ordering
  = LT
  | EQ
  | GT

instance Eq Ordering where
  LT == LT = True
  EQ == EQ = True
  GT == GT = True
  _  == _  = False

class IsString a where
  fromString ∷ Data.String.String → a

class IsNumber a where
  fromInteger ∷ GHC.Num.Integer → a

class Mul a b c | a b → c where
  (*) ∷ a → b → c

class Add a b c | a b → c where
  (+) ∷ a → b → c

class Sub a b c | a b → c where
  (-) ∷ a → b → c

class Div a b c | a b → c where
  (/) ∷ a → b → c

infixl 7 *

infixl 7 /

infixl 6 +

infixl 6 -

class Negate a where
  negate  ∷ a → a
  abs     ∷ a → a

const ∷ ∀ a b . a → b → a
const x _ = x

id ∷ ∀ a . a → a
id x = x

(.) ∷ ∀ a b c . (b → c) → (a → b) → a → c
(.) f g x = f (g x)

data IO a

instance Functor IO where
  fmap f x = x >>= (return . f)

instance Applicative IO where
  pure    = return

instance Monad IO where
  (>>=)   = rewrite "BindIO"

  (>>)    = rewrite "SeqIO"

  return  = rewrite "ReturnIO"

type List a = [a]

{-  May make sense to encapsulate common functionality between maps/sets in typeclasses. -}

data Set a

emptySet ∷ ∀ a . Set a
emptySet = rewrite "EmptySetUT"

setMember ∷ ∀ a . a → Set a → Bool
setMember = rewrite "SetMemUT"

setUpdate ∷ ∀ a . a → Bool → Set a → Set a
setUpdate = rewrite "SetUpdateUT"

setSize ∷ ∀ a . Set a → Int
setSize = rewrite "SetSizeUT"

data Map k v

emptyMap ∷ ∀ k v . (Ord k) ⇒ Map k v
emptyMap = rewrite "EmptyMapUT"

mapGet ∷ ∀ k v . (Ord k) ⇒ k → Map k v → Maybe v
mapGet = rewrite "MapGetUT"

mapMember ∷ ∀ k v . (Ord k) ⇒ k → Map k v → Bool
mapMember = rewrite "MapMemberUT"

mapUpdate ∷ ∀ k v . (Ord k) ⇒ k → Maybe v → Map k v → Map k v
mapUpdate = rewrite "MapUpdateUT"

mapSize ∷ ∀ k v . (Ord k) ⇒ Map k v → Int
mapSize = rewrite "MapSizeUT"

data Signature

checkSignature ∷ ∀ a . Key → (Signature, a) → Bool
checkSignature = rewrite "CheckSignatureUT"

data Key

data Hash

instance Eq Key
instance Ord Key

data Contract a b

type Account = Contract () ()

manager ∷ ∀ a b . Contract a b → Key
manager = rewrite "ManagerUT"

defaultAccount ∷ Hash → Account
defaultAccount = rewrite "DefaultAccountUT"

source ∷ ∀ a b . Contract a b
source = rewrite "SourceUT"

self ∷ ∀ a b . Contract a b
self = rewrite "SelfUT"

{-  c *must* be the type of the contract's storage - at the moment this is not enfored by the Haskell typesystem  -}

transferTokens ∷ ∀ a b c . a → Tez → Contract a b → c → IO (b, c)
transferTokens = rewrite "TransferTokensUT"

data Either a b
  = Left a
  | Right b

data Maybe a
  = Just a
  | Nothing

fromMaybe ∷ ∀ a . a → Maybe a → a
fromMaybe _ (Just v) = v
fromMaybe d Nothing  = d

type Pair a b = (a, b)

type Unit = ()

unit ∷ Unit
unit = ()

data Timestamp

now ∷ Timestamp
now = rewrite "NowUT"

data Nat

data Tez

amount ∷ Tez
amount = rewrite "AmountUT"

balance ∷ Tez
balance = rewrite "BalanceUT"

data Bool
  = True
  | False

otherwise ∷ Bool
otherwise = True

infixr 3 &&

(&&) ∷ Bool → Bool → Bool
(&&) = rewrite "AndUT"

infixr 2 ||

(||) ∷ Bool → Bool → Bool
(||) = rewrite "OrUT"

xor ∷ Bool → Bool → Bool
xor = rewrite "XorUT"

not ∷ Bool → Bool
not = rewrite "NotUT"

data String

hash ∷ a → String
hash = rewrite "HUT"

data Int

{-  This is *not* a side effect - however, it makes no sense without specified evaluation order, so it's in the IO monad.
    Not sure if this is useful with Juvix anyways; if the quota is static the transpiler can analyze it.  -}

stepsToQuota ∷ IO Int
stepsToQuota = rewrite "StepsToQuotaUT"

{-  Typeclass Instances   -}

{-  Equality      -}

instance Eq Tez where
  x == y = (rewrite "CmpEqUT") x y

instance Eq Int where
  x == y = (rewrite "CmpEqUT") x y

instance Eq Nat where
  x == y = (rewrite "CmpEqUT") x y

{-  Comparision   -}

instance Ord Tez where
  x <= y = (rewrite "CmpLeUT") x y

instance Ord Int where
  x <= y = (rewrite "CmpLeUT") x y

instance Ord Nat where
  x <= y = (rewrite "CmpLeUT") x y

{-  Arithmetic    -}

instance Add Int Int Int where
  (+) = rewrite "AddIntIntUT"

instance Sub Int Int Int where
  (-) = rewrite "SubIntUT"

instance Mul Int Int Int where
  (*) = rewrite "MulIntIntUT"

instance Div Int Int Int where
  (/) = rewrite "EdivIntIntUT"

instance Negate Int where
  negate  = rewrite "NegIntUT"
  abs     = rewrite "AbsIntUT"

instance Negate Nat where
  negate  = rewrite "NegNatUT"
  abs     = rewrite "AbsNatUT"

instance Add Nat Nat Nat where
  (+) = rewrite "AddNatNatUT"

instance Sub Nat Nat Nat where
  (-) = rewrite "SubNatNatUT"

instance Mul Nat Nat Nat where
  (*) = rewrite "MulNatNatUT"

instance Div Nat Nat Nat where
  (/) = rewrite "EdivNatNatUT"

instance Add Tez Tez Tez where
  (+) = rewrite "AddTezUT"

instance Sub Tez Tez Tez where
  (-) = rewrite "SubTezUT"

instance Add Int Nat Int where
  (+) = rewrite "AddIntNatUT"

instance Add Nat Int Int where
  (+) = rewrite "AddNatIntUT"

instance Mul Int Nat Int where
  (*) = rewrite "MulIntNatUT"

instance Mul Nat Int Int where
  (*) = rewrite "MulNatIntUT"

{-  Strings   -}

instance Monoid String where
  mempty  = fromString ""
  mappend = rewrite "ConcatUT"

{-  For string literal overloading.   -}

instance IsString String where
  fromString = rewrite "stringFromString"

{-  Never exposed; we need these for numeric literal overloading.   -}

instance IsNumber Tez where
  fromInteger = rewrite "tezFromInteger"
  {-# NOINLINE fromInteger #-}

instance IsNumber Int where
  fromInteger = rewrite "intFromInteger"
  {-# NOINLINE fromInteger #-}

instance IsNumber Nat where
  fromInteger = rewrite "natFromInteger"
  {-# NOINLINE fromInteger #-}

fail ∷ IO a
fail = rewrite "FailUT"

{-  Builtin library calls, rewritten by Juvix.  -}

rewrite ∷ ∀ a b . a → b
rewrite label = let x = x in x
{-# NOINLINE rewrite #-}
