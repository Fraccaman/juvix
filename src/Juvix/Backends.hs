module Juvix.Backends where

import           Juvix.Backends.Michelson
