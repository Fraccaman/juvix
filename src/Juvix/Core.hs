module Juvix.Core (
  module Juvix.Core.Dynamic,
  module Juvix.Core.Utility,
  module Juvix.Core.Types
) where

import           Juvix.Core.Dynamic
import           Juvix.Core.Types
import           Juvix.Core.Utility
