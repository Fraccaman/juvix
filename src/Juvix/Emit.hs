module Juvix.Emit where

import qualified Data.Text    as T
import           Foundation
import qualified Prelude      as P

import           Juvix.Script

emitUT ∷ ExprUT → T.Text
emitUT expr =
  case expr of

    UnitUT       → "UNIT"
    StringUT s   → T.concat ["\"", s, "\""]
    BoolUT b     → T.pack $ P.show b
    IntegerUT i  → T.pack $ P.show i
    TezUT t      → T.pack $ P.show t

    DropUT       → "DROP"
    DupUT        → "DUP"
    SwapUT       → "SWAP"
    ConstUT UnitUT        → "UNIT"
    ConstUT (BoolUT b)    → T.concat ["PUSH bool ", if b then "True" else "False"]
    ConstUT (IntegerUT i) → T.concat ["PUSH int ", T.pack (P.show i)]
    ConstUT (TezUT i)     → T.concat ["PUSH tez \"", T.pack (P.show i), "\""]
    ConstUT v             → emitUT v

    ConsPairUT   → "PAIR"
    CarUT        → "CAR"
    CdrUT        → "CDR"

    ConsSomeUT   → "SOME"
    ConsNoneUT   → "NONE"
    IfNoneUT a b → T.concat ["IF_NONE {", emitUT a, "} {", emitUT b, "}"]

    LeftUT       → "LEFT"
    RightUT      → "RIGHT"
    IfLeftUT a b → T.concat ["IF_LEFT {", emitUT a, "} {", emitUT b, "}"]

    ConsListUT   → "CONS"
    NilUT        → "NIL"
    IfConsUT a b → T.concat ["IF_CONS {", emitUT a, "} {", emitUT b, "}"]
    ListMapUT    → "MAP"
    ListReduceUT → "REDUCE"

    EmptySetUT   → "EMPTY_SET"
    SetMapUT     → "MAP"
    SetReduceUT  → "REDUCE"
    SetMemUT     → "MEM"
    SetUpdateUT  → "UPDATE"
    SetSizeUT    → "SIZE"

    EmptyMapUT   → "EMPTY_MAP"
    MapMapUT     → "MAP"
    MapReduceUT  → "REDUCE"
    MapMemUT     → "MEM"
    MapGetUT     → "GET"
    MapUpdateUT  → "UPDATE"
    MapSizeUT    → "SIZE"

    ConcatUT     → "CONCAT"

    AddSecondsToTimestampUT → "ADD"
    AddTimestampToSecondsUT → "ADD"

    AddTezUT     → "ADD"
    SubTezUT     → "SUB"
    MulTezNatUT  → "MUL"
    MulNatTezUT  → "MUL"
    EdivTezNatUT → "DIV"
    EdivTezUT    → "DIV"

    OrUT         → "OR"
    AndUT        → "AND"
    XorUT        → "XOR"
    NotUT        → "NOT"

    NegNatUT     → "NEG"
    NegIntUT     → "NEG"
    AbsNatUT     → "ABS"
    AbsIntUT     → "ABS"
    IntNatUT     → "CAST"
    AddIntIntUT  → "ADD"
    AddIntNatUT  → "ADD"
    AddNatIntUT  → "ADD"
    AddNatNatUT  → "ADD"
    SubIntUT     → "SUB"
    MulIntIntUT  → "MUL"
    MulIntNatUT  → "MUL"
    MulNatIntUT  → "MUL"
    MulNatNatUT  → "MUL"
    EdivIntIntUT → "DIV"
    EdivIntNatUT → "DIV"
    EdivNatIntUT → "DIV"
    EdivNatNatUT → "DIV"
    LslNatUT     → "LSL"
    LsrNatUT     → "LSR"
    OrNatUT      → "OR"
    AndNatUT     → "ADD"
    XorNatUT     → "XOR"
    NotNatUT     → "NOT"
    NotIntUT     → "NOT"

    SeqUT a b    → T.concat ["{", emitUT a, "; ", emitUT b, "}"]
    IfUT a b     → T.concat ["IF {", emitUT a, "} {", emitUT b, "}"]
    LoopUT a     → T.concat ["LOOP {", emitUT a, "}"]
    DipUT a      → T.concat ["DIP {", emitUT a, "}"]
    ExecUT       → "EXEC"
    LambdaUT a   → T.concat ["LAMBDA {", emitUT a, "}"]
    FailUT       → "FAIL"
    NopUT        → "NOP"

    CompareUT    → "COMPARE"
    EqUT         → "EQ"
    NeqUT        → "NEQ"
    LtUT         → "LT"
    GtUT         → "GT"
    LeUT         → "LE"
    GeUT         → "GE"

    CmpLtUT      → "CMPLT"
    CmpLeUT      → "CMPLE"
    CmpEqUT      → "CMPEQ"
    CmpGeUT      → "CMPGE"
    CmpGtUT      → "CMPGT"

    ManagerUT        → "MANAGER"
    TransferTokensUT → "TRANSFER_TOKENS"
    CreateAccountUT  → "CREATE_ACCOUNT"
    DefaultAccountUT → "DEFAULT_ACCOUNT"
    CreateContractUT → "CREATE_CONTRACT"
    NowUT            → "NOW"
    BalanceUT        → "BALANCE"
    CheckSignatureUT → "CHECK_SIGNATURE"
    HUT              → "H"
    StepsToQuotaUT   → "STEPS_TO_QUOTA"
    SourceUT         → "SOURCE"
    AmountUT         → "AMOUNT"

emitFinal ∷ ExprUT → T.Text
emitFinal expr =
  let code =
        case expr of
          NopUT → ""
          _     → emitUT expr in
  T.concat ["{", code, "}"]

emitType ∷ Type → T.Text
emitType IntT        = "int"
emitType KeyT        = "key"
emitType TezT        = "tez"
emitType BoolT       = "bool"
emitType StringT     = "string"
emitType UnitT       = "unit"
emitType (PairT a b) = T.concat ["(pair ", emitType a, " ", emitType b, ")"]
emitType (LamT _ _)  = ""
