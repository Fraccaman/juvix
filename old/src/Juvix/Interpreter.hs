{-# LANGUAGE CPP       #-}
{-# LANGUAGE MagicHash #-}

module Juvix.Interpreter where

import           Control.Concurrent
import           Control.Monad
import           Control.Monad.IO.Class
import qualified DynFlags
import           Foundation
import qualified GHC
import qualified GHC.LanguageExtensions as LangExt
import qualified GHC.Paths              as Paths
import qualified GHC.Prim               as GHC
import qualified GHCi
import qualified GHCi.RemoteTypes       as GHCi
import qualified HscMain                as GHC
import qualified Linker                 as GHC
import qualified Outputable             as GHC
import qualified Prelude                as P

import           Juvix.Core.Utility

data Command
  = Load P.String
  | Eval P.String

data Response
  = Loaded
  | Evaluated P.String

newtype InterpreterM = InterpreterM ()

ghcVersion ∷ Int
ghcVersion = __GLASGOW_HASKELL__

loadAndRun ∷ IO (Command → IO Response)
loadAndRun = do
  cmdChan ← newChan
  rspChan ← newChan
  void $ forkIO $ runInterpreter cmdChan rspChan
  return $ \cmd → do
    writeChan cmdChan cmd
    readChan rspChan

{-
runInterpreter ∷ Chan Command → Chan Response → IO a
runInterpreter cmdChan respChan = do
  oldFlags ← GHC.runGhc (Just Paths.libdir) DynFlags.getDynFlags
  let newFlags      = (DynFlags.updOptLevel 0 oldFlags) {
                          GHC.packageFlags = [DynFlags.ExposePackage "juvix" (DynFlags.PackageArg "juvix") (DynFlags.ModRenaming True [])] <> DynFlags.packageFlags oldFlags,
                          GHC.ghcMode = GHC.CompManager,
                          GHC.ghcLink = GHC.LinkInMemory,
                          GHC.hscTarget = DynFlags.HscInterpreted
                        }
      setFlags      = []
      unsetFlags    = []
      newFlags'     = foldr (flip DynFlags.wopt_set) (foldr (flip DynFlags.wopt_unset) newFlags unsetFlags) setFlags
      setGFlags     = [GHC.Opt_AutoLinkPackages]
      unsetGFlags   = []
      newFlags''    = foldr (flip DynFlags.gopt_set) (foldr (flip DynFlags.gopt_unset) newFlags' unsetGFlags) setGFlags
      setOptions    = [LangExt.UnicodeSyntax, LangExt.RebindableSyntax]
      unsetOptions  = [LangExt.ImplicitPrelude]
      finalFlags    = foldr (flip DynFlags.xopt_set) (foldr (flip DynFlags.xopt_unset) newFlags'' unsetOptions) setOptions
  hscEnv ← GHC.newHscEnv finalFlags
  forever $ do
    cmd ← liftIO (readChan cmdChan)
    resp ← case cmd of
      Eval expr → do
        fval ← GHCi.EvalThis |<< (GHCi.mkFinalizedHValue hscEnv =<< GHCi.mkRemoteRef (GHCi.HValue (GHC.unsafeCoerce# expr)))
        GHCi.EvalComplete _ (GHCi.EvalSuccess [result]) ← GHCi.evalStmt hscEnv False fval
        value ← GHCi.wormhole finalFlags result
        return (Evaluated (P.show value))
    liftIO (writeChan respChan resp)
-}

runInterpreter ∷ Chan Command → Chan Response → IO a
runInterpreter cmdChan respChan = GHC.runGhc (Just Paths.libdir) $ do
  oldFlags ← DynFlags.getDynFlags
  --mod ← GHC.findModule (GHC.mkModuleName "Juvix") Nothing
  let newFlags      = (DynFlags.updOptLevel 0 oldFlags) {
                          GHC.packageFlags = [DynFlags.ExposePackage "juvix" (DynFlags.PackageArg "juvix") (DynFlags.ModRenaming True [])] <> DynFlags.packageFlags oldFlags,
                          GHC.ghcMode = GHC.CompManager,
                          GHC.ghcLink = GHC.LinkInMemory,
                          GHC.hscTarget = DynFlags.HscInterpreted
                        }
      setFlags      = []
      unsetFlags    = []
      newFlags'     = foldr (flip DynFlags.wopt_set) (foldr (flip DynFlags.wopt_unset) newFlags unsetFlags) setFlags
      setGFlags     = [GHC.Opt_AutoLinkPackages]
      unsetGFlags   = []
      newFlags''    = foldr (flip DynFlags.gopt_set) (foldr (flip DynFlags.gopt_unset) newFlags' unsetGFlags) setGFlags
      setOptions    = [LangExt.UnicodeSyntax, LangExt.RebindableSyntax]
      unsetOptions  = [LangExt.ImplicitPrelude]
      finalFlags    = foldr (flip DynFlags.xopt_set) (foldr (flip DynFlags.xopt_unset) newFlags'' unsetOptions) setOptions
  toLink ← GHC.setSessionDynFlags finalFlags
  hscEnv ← GHC.getSession
  GHC.setTargets []
  _ ← GHC.load GHC.LoadAllTargets
  liftIO $ GHC.linkPackages hscEnv toLink

  --Just inf ← GHC.getModuleInfo mod
  let mod = GHC.mainModIs newFlags
  --liftIO $ P.print $ GHC.showSDoc newFlags $ GHC.ppr mod

  {-
  GHC.addTarget (GHC.Target (GHC.TargetModule (GHC.mkModuleName "Juvix")) True Nothing)
  GHC.depanal [] True
  GHC.load GHC.LoadAllTargets
  -}

  {-
  summary ← GHC.getModSummary (GHC.mkModuleName "Juvix")
  parsed ← GHC.parseModule summary
  typechecked ← GHC.typecheckModule parsed
  GHC.loadModule typechecked
  liftIO $ P.print $ GHC.showSDoc newFlags $ GHC.ppr mod
  -}

  --GHC.setContext [GHC.IIModule (GHC.mkModuleName "Juvix")]
  --GHC.setContext [GHC.IIModule (GHC.mkModuleName "Foundation")]
  --GHC.setContext [GHC.IIModule (GHC.mkModuleName "Prelude")]

  forever $ do
    cmd ← liftIO (readChan cmdChan)
    resp ← case cmd of
      Eval expr → do
        compiled ← GHC.dynCompileExpr expr
        return (Evaluated (P.show compiled))
    liftIO (writeChan respChan resp)
