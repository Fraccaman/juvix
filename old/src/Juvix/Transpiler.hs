module Juvix.Transpiler where

import           Control.Monad.Except
import           Control.Monad.RWS.Strict
import qualified Data.Map                         as M
import qualified Data.Text                        as T
import           Data.Typeable
import           Foundation
import qualified Prelude                          as P

import qualified Juvix.Backends.Michelson         as M hiding (Map)
import           Juvix.Core
import           Juvix.Core.CompilerTypes
import qualified Juvix.Core.GHC                   as GHC
import           Juvix.Transpiler.CoreToExpr
import           Juvix.Transpiler.CoreToType
import           Juvix.Transpiler.ExprToMichelson
import           Juvix.Transpiler.PrettyPrint
import           Juvix.Transpiler.SimplifyExpr

moduleToMichelson ∷ GHC.CoreProgram → GHC.CoreModule → ([CompileLog M.Type], Either CompileError (M.SomeExpr, M.Type, M.Type, M.Type))
moduleToMichelson l m =
  let binds ∷ GHC.CoreProgram
      binds   = l `mappend` GHC.cm_binds m
      env ∷ M.Map T.Text GHC.CoreExpr
      env     = M.fromList (P.concatMap (\case (GHC.NonRec v e) → [(prettyPrintValue v, e)]; GHC.Rec l → fmap (first prettyPrintValue) l) binds)
      computation ∷ CompilerM (M.SomeExpr, M.Type, M.Type, M.Type) M.Type
      computation = do
        core ← case filter (\case (GHC.NonRec v _) | nameToTextSimple (GHC.varName v) == "main" → True; _ → False) binds of
                  GHC.NonRec _ e:_ → tell [FrontendToCore e (GHC.exprType e)] >> return e
                  _                → throwError MainFunctionNotFound
        resolved <- resolveVars core
        expr ← coreToExpr resolved
        exprType ← coreToType core
        case exprType of
          M.LamT start@(M.PairT paramTy startStorageTy) end@(M.PairT retTy endStorageTy) | startStorageTy == endStorageTy → do
            case (M.liftType paramTy, M.liftType startStorageTy, M.liftType retTy, M.liftType endStorageTy) of
              (DynamicType (Proxy ∷ Proxy paramTyLifted), DynamicType (Proxy ∷ Proxy startStorageTyLifted), DynamicType (Proxy ∷ Proxy retTyLifted), DynamicType (Proxy ∷ Proxy endStorageTyLifted)) → do
                simplified ← simplifyExpr expr
                michelson ← exprToMichelson simplified
                (M.SomeExpr (expr ∷ M.Expr (M.Stack a) (M.Stack b)), _) ← M.liftUntyped michelson (M.typeToStack start) (DynamicType (Proxy :: Proxy startStorageTyLifted))
                case (eqT ∷ Maybe (a :~: (M.Pair paramTyLifted startStorageTyLifted, ())), eqT ∷ Maybe (b :~: (M.Pair retTyLifted endStorageTyLifted, ()))) of
                  (Just Refl, Just Refl) → do
                    optimized ← M.optimize expr
                    return (M.SomeExpr optimized, paramTy, retTy, startStorageTy)
                  _ → do
                    throwError (NotYetImplemented (T.concat ["Cannot unify start/end stack types: start ", prettyPrintValue start, " / end ", prettyPrintValue end, " with compilation output type ", T.pack (P.show (typeOf (undefined ∷ b))), "; this is a bug in Juvix and should be reported"]))
          _ → throwError (NotYetImplemented (T.concat ["Invalid type for main function - must be ∷ (param, storage) → (ret, storage) but was instead ", prettyPrintValue exprType]))
      (res, _, logs) = runRWS (runExceptT computation) (Env env) [FuncResult]
  in (logs, res)
