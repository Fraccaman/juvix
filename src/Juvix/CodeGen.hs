module Juvix.CodeGen where

import           Control.Monad.Except
import           Control.Monad.Writer
import           Protolude

import           Idris.AbsSyntax
import           Idris.Core.TT            hiding (Name)
import           Idris.ElabDecls
import           Idris.Main
import           Idris.Options
import           IRTS.CodegenCommon
import           IRTS.Compiler
import           IRTS.Lang

import qualified Juvix.Backends.Michelson as M
import           Juvix.Lang
import           Juvix.Utility

data Opts = Opts {inputs :: [FilePath],
                  output :: FilePath }

showUsage ∷ IO ()
showUsage = do putText "A code generator which is intended to be called by the compiler, not by a user."
               putText "Usage: idris-codegen-juvix <ibc-files> [-o <output-file>]"
               exitWith ExitSuccess

getOpts ∷ IO Opts
getOpts = do xs <- getArgs
             return $ process (Opts [] "main.jvx") xs
  where
    process opts ("-o":o:xs) = process (opts { output = o }) xs
    process opts (x:xs)      = process (opts { inputs = x:inputs opts }) xs
    process opts []          = opts

codeGenSdecls ∷ CodeGenerator
codeGenSdecls ci = do
  putText $ "Output file : " <> show (outputFile ci)
  let decls = liftDecls ci
  putText $ "Number of decls: " <> show (length decls)
  let main = findMain decls
  putText $ "Main decl print: " <> show main
  let decl = snd main
  putText $ "Main decl prettified: " <> prettyPrintValue decl
  let LFun _ _ args body = decl
      expr = LLam args body
  putText $ "Main expr prettified: " <> prettyPrintValue expr
  case runWriter $ runExceptT (M.compileToMichelsonSourceFile expr ∷ (ExceptT M.CompilationError (Writer [M.CompilationLog]) Text)) of
    (res, logs) -> do
      mapM_ (putText . prettyPrintValue) logs
      case res of
        Right output -> do
          writeFile (outputFile ci) output
          putText "Compilation success!"
        Left err -> do
          putText $ "Error during transpilation: " <> prettyPrintValue err
          exitFailure

findMain ∷ [(Name, LDecl)] → (Name, LDecl)
findMain decls = let Just f = head $ filter (\(name, _) -> name == NS (UN "main") ["Main"]) decls in f

sdeclMain ∷ Opts → Idris ()
sdeclMain opts = do elabPrims
                    _ <- loadInputs (inputs opts) Nothing
                    mainProg <- elabMain
                    ir <- compile (Via IBCFormat "juvix") (output opts) (Just mainProg)
                    runIO $ codeGenSdecls ir

main ∷ IO ()
main = do opts <- getOpts
          if (null (inputs opts))
             then showUsage
             else runMain (sdeclMain opts)

{-
data CodegenInfo = CodegenInfo {
    outputFile    :: String
  , outputType    :: OutputType
  , targetTriple  :: String
  , targetCPU     :: String
  , includes      :: [FilePath]
  , importDirs    :: [FilePath]
  , compileObjs   :: [String]
  , compileLibs   :: [String]
  , compilerFlags :: [String]
  , debugLevel    :: DbgLevel
  , simpleDecls   :: [(Name, SDecl)] -- most low level
  , defunDecls    :: [(Name, DDecl)]
  , liftDecls     :: [(Name, LDecl)]
  , interfaces    :: Bool
  , exportDecls   :: [ExportIFace]
  , ttDecls       :: [(Name, TTDecl)]
  }
-}
