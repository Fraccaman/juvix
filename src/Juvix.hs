module Juvix (
  compileToTz,
  pprint
) where

import           Control.Monad
import qualified Data.Text        as T
import qualified Data.Text.IO     as T
import           Foundation
import qualified GHC
import qualified Prelude          as P

import           Juvix.GHC
import qualified Juvix.Michelson  as M
import           Juvix.Transpiler
import           Juvix.Types
import           Juvix.Utility

compileToTz ∷ P.FilePath → Bool → IO (Either CompileError T.Text)
compileToTz fn log = do
  lib ← compileModule "Juvix/Lib.hs"
  mod ← compileModule fn
  let (logs, compiled) = moduleToMichelson (GHC.cm_binds lib) mod
  when log (P.mapM_ (T.putStrLn . pprint) logs)
  return (do
    (M.SomeExpr code, paramTy, retTy, storageTy) ← compiled
    return $ T.unlines [
      T.concat ["parameter ", M.emitType paramTy, ";"],
      T.concat ["return ", M.emitType retTy, ";"],
      T.concat ["storage ", M.emitType storageTy, ";"],
      T.concat ["code ", M.emitFinal code, ";"]
      ])
