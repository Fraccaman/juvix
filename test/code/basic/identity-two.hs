module Identity where

import           Juvix.Lib

-- This identity example is *not* fully optimized by Juvix (yet).

main âˆ· (String, ()) -> (String, ())
main (str, ()) = (str, ())
