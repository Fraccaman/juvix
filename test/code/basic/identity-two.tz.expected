parameter string;
return string;
storage unit;
code {{{{DUP; {{{{{DUP; CDR}; SWAP}; CAR}; {UNIT; {{{SWAP; DUP}; DIP {SWAP}}; PAIR}}}; {{SWAP; DROP}; {SWAP; DROP}}}}; SWAP}; DROP}};
